File: languages/java/src/main/java/com/bloock/sdk/entity/identity/CredentialOfferBuilder.java
package com.bloock.sdk.entity.identity;

import com.bloock.sdk.bridge.Bridge;
import com.bloock.sdk.bridge.proto.Config;
import com.bloock.sdk.bridge.proto.Identity;
import com.bloock.sdk.bridge.proto.IdentityEntities;
import com.bloock.sdk.bridge.proto.Shared;

import java.util.ArrayList;
import java.util.List;

public class CredentialOfferBuilder {
    private String schemaId;
    private String holderKey;
    private Config.ConfigData configData;

    private List<IdentityEntities.BooleanAttribute> booleanAttributes;
    private List<IdentityEntities.DateAttribute> dateAttributes;
    private List<IdentityEntities.DateTimeAttribute> datetimeAttributes;
    private List<IdentityEntities.MultiChoiceAttribute> multichoiceAttributes;
    private List<IdentityEntities.NumberAttribute> numberAttributes;

    public CredentialOfferBuilder(String schemaId, String holderKey, Config.ConfigData configData) {
        this.schemaId = schemaId;
        this.holderKey = holderKey;
        this.configData = configData;

        this.booleanAttributes = new ArrayList<>();
        this.dateAttributes = new ArrayList<>();
        this.datetimeAttributes = new ArrayList<>();
        this.multichoiceAttributes = new ArrayList<>();
        this.numberAttributes = new ArrayList<>();
    }

    public CredentialOfferBuilder withBooleanAttribute(String key, boolean value) {
        this.booleanAttributes.add(new BooleanAttribute(key, value).toProto());
        return this;
    }

    public CredentialOfferBuilder withDateAttribute(String key, Long value) {
        this.dateAttributes.add(new DateAttribute(key, value).toProto());
        return this;
    }

    public CredentialOfferBuilder withDatetimeAttribute(String key, Long value) {
        this.datetimeAttributes.add(new DatetimeAttribute(key, value).toProto());
        return this;
    }

    public CredentialOfferBuilder withMultichoiceAttribute(String key, String value) {
        this.multichoiceAttributes.add(new MultichoiceAttribute(key, value).toProto());
        return this;
    }

    public CredentialOfferBuilder withNumberAttribute(String key, Long value) {
        this.numberAttributes.add(new NumberAttribute(key, value).toProto());
        return this;
    }

    public CredentialOffer build() throws Exception {
        Bridge bridge = new Bridge();

        Identity.CreateCredentialOfferRequest req = Identity.CreateCredentialOfferRequest.newBuilder()
                .setConfigData(this.configData)
                .setSchemaId(this.schemaId)
                .setHolderKey(this.holderKey)
                .addAllBooleanAttributes(this.booleanAttributes)
                .addAllDateAttributes(this.dateAttributes)
                .addAllDatetimeAttributes(this.datetimeAttributes)
                .addAllMultichoiceAttributes(this.multichoiceAttributes)
                .addAllNumberAttributes(this.numberAttributes)
                .build();

        Identity.CreateCredentialOfferResponse response = bridge.getIdentity().createCredentialOffer(req);

        if (response.getError() != Shared.Error.getDefaultInstance()) {
            throw new Exception(response.getError().getMessage());
        }

        return CredentialOffer.fromProto(response.getCredentialOffer());
    }
}

File: languages/php/Bloock/Entity/Identity/CredentialOfferBuilder.php
<?php

namespace Bloock\Entity\Identity;

use Bloock\Bridge\Bridge;
use Bloock\ConfigData;
use Bloock\CreateCredentialOfferRequest;
use Exception;

class CredentialOfferBuilder
{
    private string $schemaId;
    private string $holderKey;
    private ConfigData $configData;

    private array $booleanAttributes;
    private array $dateAttributes;
    private array $datetimeAttributes;
    private array $multichoiceAttributes;
    private array $numberAttributes;

    public function __construct(string $schemaId, string $holderKey, ConfigData $configData)
    {
        $this->schemaId = $schemaId;
        $this->holderKey = $holderKey;
        $this->configData = $configData;

        $this->booleanAttributes = [];
        $this->dateAttributes = [];
        $this->datetimeAttributes = [];
        $this->multichoiceAttributes = [];
        $this->numberAttributes = [];
    }

    public function withBooleanAttribute(string $key, bool $value): CredentialOfferBuilder
    {
        $attribute = new BooleanAttribute($key, $value);
        $this->booleanAttributes[] = $attribute->toProto();
        return $this;
    }

    public function withDateAttribute(string $key, int $value): CredentialOfferBuilder
    {
        $attribute = new DateAttribute($key, $value);
        $this->dateAttributes[] = $attribute->toProto();
        return $this;
    }

    public function withDatetimeAttribute(string $key, int $value): CredentialOfferBuilder
    {
        $attribute = new DatetimeAttribute($key, $value);
        $this->datetimeAttributes[] = $attribute->toProto();
        return $this;
    }

    public function withMultichoiceAttribute(string $key, string $value): CredentialOfferBuilder
    {
        $attribute = new MultichoiceAttribute($key, $value);
        $this->multichoiceAttributes[] = $attribute->toProto();
        return $this;
    }

    public function withNumberAttribute(string $key, int $value): CredentialOfferBuilder
    {
        $attribute = new NumberAttribute($key, $value);
        $this->numberAttributes[] = $attribute->toProto();
        return $this;
    }

    public function build(): CredentialOffer
    {
        $bridge = new Bridge();

        $req = new CreateCredentialOfferRequest();
        $req->setConfigData($this->configData);
        $req->setSchemaId($this->schemaId);
        $req->setHolderKey($this->holderKey);
        $req->setBooleanAttributes($this->booleanAttributes);
        $req->setDateAttributes($this->dateAttributes);
        $req->setDatetimeAttributes($this->datetimeAttributes);
        $req->setMultichoiceAttributes($this->multichoiceAttributes);
        $req->setNumberAttributes($this->numberAttributes);

        $res = $bridge->identity->CreateCredentialOffer($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return CredentialOffer::fromProto($res->getCredentialOffer());
    }
}